cpp/src/round/round.cu

```diff
@@ -47,102 +52,190 @@ T __device__ generic_abs(T value)
 }
 
 template <typename T, typename std::enable_if_t<std::is_signed<T>::value>* = nullptr>
-bool __device__ is_negative(T value)
+int16_t __device__ generate_sign(T value)
 {
-  return value < 0;
+  return value > 0 ? 1 : value < 0 ? -1 : 0;
 }
 
 // this is needed to suppress warning: pointless comparison of unsigned integer with zero
 template <typename T, typename std::enable_if_t<not std::is_signed<T>::value>* = nullptr>
-bool __device__ is_negative(T value)
+int16_t __device__ generate_sign(T value)
 {
-  return false;
+  return value > 0 ? 1 : 0;
 }
 
-struct round_fn {
-  template <typename T, typename... Args>
-  std::enable_if_t<not cudf::is_numeric<T>(), std::unique_ptr<column>> operator()(Args&&... args)
+template <typename T>
+constexpr inline auto is_supported_round_type()
+{
+  return cudf::is_numeric<T>() && not std::is_same<T, bool>::value;
+}
+
+template <typename T>
+struct half_up_zero {
+  T n;  // unused in the decimal_places = 0 case
+  template <typename U = T, typename std::enable_if_t<cudf::is_floating_point<U>()>* = nullptr>
```

Comment:
Why do you need a type alias here? Can't you just use T?

Reply:
I will give it a shot, I thought it was necessary.

Reply:
Yea, I remember this being a pattern that was necessary. If you remove the `U = T` you get:
```
error: declaration of template parameter 'T' shadows template parameter
```

Reply:
But why do you need a type parameter for the operator() and also for the struct itself?  Can't you just do 

```
template <typename std::enable_if_t<cudf::is_floating_point<T>()>* = nullptr>
__device__ U operator()(U e) { ... }
```

Reply:
Then you get:
```
error: class "std::enable_if<false, void>" has no member "type"
```
I looked into it and it is because:
> SFINAE only works for deduced template arguments. In your case your method did not depend on any parameter from the method call so it is not in a deduced context. Everything is already known while instantiate the class itself.

https://www.godbolt.org/z/sqT6EY
https://stackoverflow.com/questions/50713692/why-enable-if-cannot-be-used-to-disable-this-declaration-here/50719848

Reply:
Yeah this is a common trick to SFINAE out member functions that don't otherwise have template parameters.

Reply:
That said, do these functors need to be templates? Or can the `operator()` just be a template?

```
struct half_up_zero{

   template <typename T, typename std::enable_if_t<cudf::is_floating_point<T>()>* = nullptr>
   T operator()(T e){...}

};
```

};

Reply:
Yep, we need if for the `T n` member of the function objects.